tests/cupy_tests/cuda_tests/test_cub.py

```diff
@@ -0,0 +1,159 @@
+import functools
+import unittest
+
+import cupy
+from cupy import testing
+
+
+if cupy.cuda.cub_enabled is False:
+    raise unittest.SkipTest("The CUB module is not built. Skip the tests.")
+
+
+def for_contiguous_axes(name='axis'):
+    """Decorator for parametrizing tests with possible contiguous axes.
+
+    Args:
+        axis(str): Argument name to which specified axis are passed.
+
+    Adapted from tests/cupy_tests/fft_tests/test_fft.py.
+    """
+    def decorator(impl):
+        @functools.wraps(impl)
+        def test_func(self, *args, **kw):
+            ndim = len(self.shape)
+            for i in range(ndim):
+                a = ()
+                for j in range(ndim-1, i-1, -1):
+                    a = (j,) + a
+                try:
+                    kw[name] = a
+                    impl(self, *args, **kw)
+                except Exception:
+                    print(name, 'is', a, ', ndim is', ndim, ', shape is',
+                          self.shape)
+                    raise
+        return test_func
+    return decorator
+
+
+def timing(test, func, arr, axis, runs=20):
+    cupy.cuda.cub_enabled = True
+    t_cub = 0.
+    for i in range(runs):
+        test.start.record()
+        getattr(arr, func)(axis=axis)
+        test.stop.record()
+        test.stop.synchronize()
+        t_cub += cupy.cuda.get_elapsed_time(test.start, test.stop)
+
+    cupy.cuda.cub_enabled = False
+    t_cupy = 0.
+    for i in range(runs):
+        test.start.record()
+        getattr(arr, func)(axis=axis)
+        test.stop.record()
+        test.stop.synchronize()
+        t_cupy += cupy.cuda.get_elapsed_time(test.start, test.stop)
+
+    # TODO(leofang): raise PerformanceWarning here?
+    print("CUB: {:10.5f}; CuPy: {:10.5f} (ms), for {} runs, shape={}, axis={},"
+          " dtype={}".format(t_cub, t_cupy, runs, arr.shape, axis, arr.dtype))
+    cupy.cuda.cub_enabled = True  # restore
+
+
+# This class compares CUB results against NumPy's
+@testing.parameterize(*testing.product({
+    'shape': [(10,), (10, 20), (10, 20, 30), (10, 20, 30, 40)],
+}))
+@testing.gpu
+class TestCUBreduction(unittest.TestCase):
```

Comment:
`cupy_tests/cuda_tests/test_cub.py` should be a test for `cupy/cuda/cub.pyx`.
Function tests should be in a test file of each function. (maybe it is better to parameterize with {"cub_enabled: [True, False]})

Reply:
I had hesitation about this choice (putting all CUB-related tests in one place). The reason for doing this is that directly testing `cupy.cuda.cub` requires a bit more work. If you think it’s best to scatter the tests into existing modules with parametrization, I’ll follow.

Reply:
As for testing `cupy.cuda.cub` itself, currently other wrappers are not much tested as well, so it is an acceptable option to leave it as is. It is better to cover them of course, but the priority is *relatively* lower as we assume that (1) end-to-end test (calling functions that use wrappers and comparing the result with NumPy) should be able to find bugs in wrappers and (2) bugs are less likely to happen as the code is simple.
So, yes, please test them in tests of each function.

Reply:
OK. Just sent the CUB tests to the corresponding test modules. I hope I got everything right. PTAL.